src/python/pants/build_graph/address.py

```diff
@@ -171,10 +179,12 @@ def file_to_address(self) -> "Address":
 
         expected_prefix = f"..{os.path.sep}" * parent_count
         if self.target_component[: self.target_component.rfind(os.path.sep) + 1] != expected_prefix:
-            raise InvalidTargetName(
-                "A target may only be defined in a directory containing a file that it owns in "
-                f"the filesystem: `{self.target_component}` is not at-or-above the file "
-                f"`{self.path_component}`."
+            raise AmbiguousTargetName(
+                "The target component `{self.target_component}` is ambiguous, because it might "
+                "represent either a target with a slash (`/`) in the name, or a target nested "
+                "inside of a directory.\n\nIf there is a target with a slash in its name, "
+                "consider replacing the slash with another separator character like `_`, or "
+                "moving the target closer to the file(s) that it owns."
```

Comment:
Does moving the target closer to the file solve the problem if the target name still has a slash? 

I think the "consider replacing" language is too passive. "Please replace" would be a better way of saving people (users and developers) headaches going forward.

Reply:
The issue is that the case is ambiguous: as the error message says, we cannot know whether the target name contains a slash, or whether it's referring to a directory. We cannot say definitively from this position which it is.

But upon reflection, although this is where the error is currently exposed, I think that this is probably the wrong place to have this error. A better spot would be where we actually instantiate the target. That will also be easier to backport.